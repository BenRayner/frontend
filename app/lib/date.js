import moment from 'moment';

export function getDurationString(from, to = moment(), format = 'full') {
  if (getDurationString.formats.indexOf(format) === -1) {
    throw new Error(`getDurationString: Unknown format \`${format}\`.`);
  }

  const diff = moment(to).diff(from);

  const duration = moment.duration(diff);

  const hours = duration.get('hours');
  const minutes = duration.get('minutes');
  const seconds = duration.get('seconds');

  let times = [];

  if (hours > 0) {
    times.push([hours, "hour"]);
  }

  if (minutes > 0) {
    times.push([minutes, "minute"]);
  }

  if (seconds > 0) {
    times.push([seconds, "second"]);
  }

  if (times.length === 0) {
    times.push([0, "second"]);
  }

  if (format === "short") {
    times = times.slice(0, 2);
    // TODO: Rount next significant span
  }

  if (format === "micro") {
    times = times.slice(0, 1);
    // TODO: Rount next significant span
  }

  return times
    .map(([amount, label]) => {
      if (format === "full") {
        return `${amount} ${label}${amount === 1 ? '' : 's'}`;
      } else {
        return `${amount}${label[0]}`;
      }
    })
    .join(format === "full" ? ', ' : ' ');
}

getDurationString.formats = [
  'full',
  'short',
  'micro'
];
