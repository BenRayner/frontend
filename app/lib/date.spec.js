/* global describe, it, expect */
import MockDate from 'mockdate';
import { getDurationString } from './date';

describe('getDurationString', () => {
  describe('formats', () => {
    it('are supplied as an array', () => {
      expect(getDurationString.formats).toBeInstanceOf(Array);
      expect(getDurationString.formats.length).toMatchSnapshot();
    });
  });

  it('defaults to full dates', () => {
    expect(getDurationString("2016-10-05T03:39:57.000Z", "2016-10-05T03:40:02.000Z")).toMatchSnapshot();
  });

  getDurationString.formats.map((format) => {
    it(`correctly handles \`${format}\` dates`, () => {
      expect(getDurationString("2016-10-05T03:39:57.000Z", "2016-10-05T03:40:02.000Z", format)).toMatchSnapshot();
      expect(getDurationString("2016-09-15T06:49:27.000Z", "2016-09-15T06:52:55.000Z", format)).toMatchSnapshot();
      expect(getDurationString("2016-09-15T06:49:27.000Z", "2016-09-15T08:52:55.000Z", format)).toMatchSnapshot();
    });
  });

  it('falls back to `now` when not supplied a `to` value', () => {
    MockDate.set("2016-10-09T08:10:25.000Z");
    expect(getDurationString("2016-10-05T03:40:02.000Z")).toMatchSnapshot();
    expect(getDurationString("2016-10-05T03:40:02.000Z", undefined, "short")).toMatchSnapshot();
    MockDate.reset();
  });
});
