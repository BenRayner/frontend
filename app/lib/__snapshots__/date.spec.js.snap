exports[`getDurationString correctly handles \`full\` dates 1`] = `"5 seconds"`;

exports[`getDurationString correctly handles \`full\` dates 2`] = `"3 minutes, 28 seconds"`;

exports[`getDurationString correctly handles \`full\` dates 3`] = `"2 hours, 3 minutes, 28 seconds"`;

exports[`getDurationString correctly handles \`micro\` dates 1`] = `"5s"`;

exports[`getDurationString correctly handles \`micro\` dates 2`] = `"3m"`;

exports[`getDurationString correctly handles \`micro\` dates 3`] = `"2h"`;

exports[`getDurationString correctly handles \`short\` dates 1`] = `"5s"`;

exports[`getDurationString correctly handles \`short\` dates 2`] = `"3m 28s"`;

exports[`getDurationString correctly handles \`short\` dates 3`] = `"2h 3m"`;

exports[`getDurationString defaults to full dates 1`] = `"5 seconds"`;

exports[`getDurationString falls back to \`now\` when not supplied a \`to\` value 1`] = `"4 hours, 30 minutes, 23 seconds"`;

exports[`getDurationString falls back to \`now\` when not supplied a \`to\` value 2`] = `"4h 30m"`;

exports[`getDurationString formats are supplied as an array 1`] = `3`;
